### ISSUE TITLE
Possibly a false positive warning W0640: Cell variable category defined in loop (cell-var-from-loop)

### ISSUE THREAD

**Issue by kowalcj0** [PyCQA/pylint#5508]
### Bug description

Hi,

I'm little confused by this warning which `pylint 2.12.2` found in the following `example.py`:

```python
"""cell-var-from-loop example."""

grouped_messages = {
    "category_a": [
        {"key_a": 3},
        {"key_a": 1},
        {"key_a": 2},
    ],
    "category_b": [
        {"key_b": 5},
        {"key_b": 3},
        {"key_b": 8},
    ],
}
sorting_keys = {
    "category_a": "key_a",
    "category_b": "key_b",
}

sorted_messages = {}

for category, messages in grouped_messages.items():
    sorted_messages[category] = sorted(
        (message for message in messages),
        key=lambda msg: msg[sorting_keys[category]]
    )

print(sorted_messages)
```

I got rid of this warning by replacing `key=lambda msg: msg[sorting_keys[category]]` with `key=operator.itemgetter(sorting_keys[category])`.
TBH I'm not sure why pylint complains about `category` being defined in loop.
Especially, because both versions of the code, give exactly the same result:

```shell
python3 example.py 
{'category_a': [{'key_a': 1}, {'key_a': 2}, {'key_a': 3}], 'category_b': [{'key_b': 3}, {'key_b': 5}, {'key_b': 8}]}
```

This might be a continuation of: https://github.com/PyCQA/pylint/pull/4827

Thanks

### Command used

```shell
pylint example.py
```

### Pylint output

```shell
************* Module example
example.py:25:41: W0640: Cell variable category defined in loop (cell-var-from-loop)

------------------------------------------------------------------
Your code has been rated at 8.33/10 (previous run: 8.33/10, +0.00)
```

### Expected behavior

No warnings

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.0
Python 3.10.0 (default, Oct  4 2021, 00:00:00) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)
```

### OS / Environment

Fedora 35

### Additional dependencies

astroid==2.9.0
isort==5.10.1
lazy-object-proxy==1.6.0
mccabe==0.6.1
platformdirs==2.4.0
pylint==2.12.2
toml==0.10.2
wrapt==1.13.3

**Comment by jacobtylerwalls** [comment]
Reproduced @ 27e72725fa8c216cb18ac833c20c658ec05ab9f5.

We try to exclude lambdas that are called immediately, but we're fooled by the violation of the assumption of direct parentage here:

https://github.com/pylint-dev/pylint/blob/51c28bc922dfb8da42e680ed7190eb942889df01/pylint/checkers/utils.py#L447-L449

The lambda's parent is a `Keyword` node, and _that_ has a parent of `Call`.
